### ISSUE TITLE
Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied

### ISSUE THREAD

**Issue by wzhsh90** [dosgo/ngrok-android#3]
Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied

**Comment by dosgo** [comment]
It should be that Android 9 or higher does not run in this way, and it is packaged with api 23

**Comment by dosgo** [comment]
Or you can find another operation plan, there should be a way

**Comment by wzhsh90** [comment]
My Android version is 10.I changed the compile version. only use 23 ?




&nbsp;compileSdkVersion 30


&nbsp; &nbsp; defaultConfig {
&nbsp; &nbsp; &nbsp; &nbsp; applicationId "com.a16v16.ngrok_c"
&nbsp; &nbsp; &nbsp; &nbsp; minSdkVersion 15
&nbsp; &nbsp; &nbsp; &nbsp; targetSdkVersion 30
&nbsp; &nbsp; &nbsp; &nbsp; versionCode 1
&nbsp; &nbsp; &nbsp; &nbsp; versionName "1.0"
&nbsp; &nbsp; }





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午4:06
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
It should be that Android 9 or higher does not run in this way, and it is packaged with api 23

**Comment by dosgo** [comment]
There should be no such problem with 23, in order to maintain backward compatibility

**Comment by wzhsh90** [comment]
Ok, I rollback to test




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午4:12
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
There should be no such problem with 23, in order to maintain backward compatibility

**Comment by wzhsh90** [comment]
I rollback to 23


Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=26, Text file busy




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午4:12
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
There should be no such problem with 23, in order to maintain backward compatibility

**Comment by wzhsh90** [comment]
my purpose to use ngrok on android . use app as a server.


Another solution&nbsp; is install termux and then&nbsp; install ngrok on it. this solution is ok but&nbsp; not I want.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午4:12
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
There should be no such problem with 23, in order to maintain backward compatibility

**Comment by dosgo** [comment]
It has not been maintained for a long time, you can consider using https://github.com/dosgo/ngrok-java version, but there are bugs. . You can slightly modify

**Comment by wzhsh90** [comment]
can I use it for android ?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午4:28
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
It has not been maintained for a long time, you can consider using https://github.com/dosgo/ngrok-java version, but there are bugs. . You can slightly modify

**Comment by dosgo** [comment]
Of course, the main development language for Android is not java? I wrote such a move when there was no java version at the beginning, and then I wrote java.

**Comment by wzhsh90** [comment]
Are you test it on Android ?


Java can write for Android , but some libs can not use it on Android.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午4:31
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
Of course, the main development language for Android is not java? I wrote such a move when there was no java version at the beginning, and then I wrote java.

**Comment by dosgo** [comment]
And the java version logic is clear, if you have a bug, you can fix it.

**Comment by dosgo** [comment]
Most libraries can be used on android.

**Comment by dosgo** [comment]
This dependency is very small, basically using the tls function. . .

**Comment by wzhsh90** [comment]
Are you test ?


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午4:34
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
Most libraries can be used on android.

**Comment by dosgo** [comment]
I don't remember, it's been a long time

**Comment by dosgo** [comment]
Can it be used?

**Comment by wzhsh90** [comment]
I will try it later.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dosgo/ngrok-android"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月9日(星期五) 下午5:06
***@***.***&gt;;
抄送:&nbsp;"@***@***.******@***.***&gt;;
主题:&nbsp;Re: [dosgo/ngrok-android] Cannot run program "/data/user/0/com.a16v16.ngrok_c/files/ngrokc": error=13, Permission denied (#3)





 
Can it be used?
